### PR TITLE
KFSPTS-14051 gracefully handle bad email addresses in pdp send ach notification job

### PR DESCRIPTION
**Pull request by jhulslander**

### CODE REVIEW

File: src/main/java/edu/cornell/kfs/pdp/CUPdpKeyConstants.java [hunk 1]
```diff
@@ -27,5 +27,9 @@
 	public static final String MESSAGE_PDP_ACH_ADVICE_EMAIL_ATTACHMENT_PAYMENT_TABLE_ITEM_LINE = "message.pdp.achAdvice.email.attachment.paymentTableItemLine";
 	
 	public static final String ERROR_PDP_PAYMENTLOAD_INVALID_EMPLOYEE_ID = "error.pdp.paymentLoad.invalidEmployeeId";
+
```
Comment (ajd299) [inline src/main/java/edu/cornell/kfs/pdp/CUPdpKeyConstants.java]: I noticed an extra semicolon in unchanged code. I assume we do not want to address it as part of this commit, but what are the next steps to removing it?

Comment (ajd299) [reply src/main/java/edu/cornell/kfs/pdp/CUPdpKeyConstants.java]: public static final String MESSAGE_PDP_ACH_ADVICE_EMAIL_BODY_PAYMENT_HEADER_LINE_ONE = "message.pdp.achAdvice.email.body.paymentHeaderLineOne";;

File: src/main/java/edu/cornell/kfs/pdp/batch/PDPBadEmailRecord.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    public String getEmailAddress() {
+        return emailAddress;
+    }
+
+    public String getDisbursementNumber() {
+        return disbursementNumber;
+    }
+
+    public void logBadEmailRecord() {
+        LOG.error("logBadEmailRecord, payeeIdL '" + payeeId + "' payment group ID: '" + paymentGroupId + "' email address: '" +
```
Comment (ajd299) [inline src/main/java/edu/cornell/kfs/pdp/batch/PDPBadEmailRecord.java]: Is payeeIdL a typo or intentional?

File: src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationServiceImpl.java [hunk 3]
```diff
@@ -186,4 +227,10 @@ public void setAchBundlerHelperService(AchBundlerHelperService achBundlerHelperS
     public void setAchBundlerAdviceDao(AchBundlerAdviceDao achBundlerAdviceDao) {
         this.achBundlerAdviceDao = achBundlerAdviceDao;
     }
+    
+    @NonTransactional
+    public void setCuAchAdviceNotificationWrrorReportService(
```
Comment (ajd299) [inline src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationServiceImpl.java]: Typo: Wrror => Error

File: src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationServiceImpl.java [hunk 4]
```diff
@@ -117,22 +131,49 @@ public void sendAdviceNotifications() {
...
+        PDPBadEmailRecord badEmailRecord = new PDPBadEmailRecord(paymentGroup.getPayeeId(), paymentGroup.getId(), paymentGroup.getAdviceEmailAddress(), 
+                paymentGroup.getDisbursementNbr());
+        badEmailRecord.logBadEmailRecord();
+        badEmailRecords.add(badEmailRecord);
+    }
+    
+    @NonTransactional
+    private void createAndEmailErrorReport(List<PDPBadEmailRecord> badEmailRecords) {
+        if (CollectionUtils.isNotEmpty(badEmailRecords)) {
+            LOG.info("createBadEmailReport, there are " + badEmailRecords.size() + " bad email records to report");
```
Comment (ajd299) [inline src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationServiceImpl.java]: should records be changed to addresses as it is in the else block below?

File: src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationServiceImpl.java [hunk 5]
```diff
@@ -15,38 +15,47 @@
...
+import edu.cornell.kfs.pdp.batch.PDPBadEmailRecord;
+import edu.cornell.kfs.pdp.batch.service.CuAchAdviceNotificationWrrorReportService;
 import edu.cornell.kfs.pdp.dataaccess.AchBundlerAdviceDao;
 import edu.cornell.kfs.pdp.service.CuPdpEmailService;
 
 /**
  * @see org.kuali.kfs.pdp.batch.service.AchAdviceNotificationService
  */
 public class CuAchAdviceNotificationServiceImpl implements AchAdviceNotificationService {
+    private static final Logger LOG = LogManager.getLogger(CuAchAdviceNotificationWrrorReportServiceImpl.class);
```
Comment (ajd299) [inline src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationServiceImpl.java]: Is CuAchAdviceNotificationWrrorReportServiceImpl correct? Should it match the current class name?

File: src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationWrrorReportServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    @Override
+    public File createBadEmailReport(List<PDPBadEmailRecord> badEmailRecords) {
+        LOG.info("createBadEmailReport, there are " + badEmailRecords.size() + " bad email records to report");
+        initiatlizeErrorReport();
+        printErrorReportDetails(badEmailRecords);
+        reportWriterService.destroy();
+        return reportWriterService.getReportFile();
+    }
+    
+    private void initiatlizeErrorReport() {
```
Comment (ajd299) [inline src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationWrrorReportServiceImpl.java]: typo initiatlize => initialize

File: src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationWrrorReportServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+    private List<String> findPDPToEmailAddress() {
+        Collection addresses = parameterService.getParameterValuesAsString(KFSConstants.ParameterNamespaces.PDP, KfsParameterConstants.BATCH_COMPONENT, 
+                CUPdpParameterConstants.PDP_ACH_INVALID_EMAIL_ERROR_REPORT_TO_ADDRESSES);
+        return new ArrayList<String>(addresses);
+    }
+    
+    @Override
+    public void validateEmailAddress(String email) throws AddressException {
+        if (email == null) {
+            throw new AddressException("The email must not be null/");
```
Comment (ajd299) [inline src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationWrrorReportServiceImpl.java]: Is the slash at the end intentional?

File: src/test/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationWrrorReportServiceImplTest.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+import javax.mail.internet.AddressException;
+
+import org.apache.commons.lang.StringUtils;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.kuali.kfs.sys.KFSConstants;
+
+public class CuAchAdviceNotificationWrrorReportServiceImplTest {
```
Comment (ajd299) [inline src/test/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationWrrorReportServiceImplTest.java]: typo Wrror => Error

File: src/main/java/edu/cornell/kfs/pdp/CUPdpKeyConstants.java (hunk 1 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/pdp/CUPdpKeyConstants.java]: easy enough to change, I shall do it in this PR

File: src/main/java/edu/cornell/kfs/pdp/batch/PDPBadEmailRecord.java (hunk 2 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/pdp/batch/PDPBadEmailRecord.java]: typo, I shall fix

File: src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationServiceImpl.java (hunk 5 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationServiceImpl.java]: good catch, needs fixed

File: src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationWrrorReportServiceImpl.java (hunk 6 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationWrrorReportServiceImpl.java]: fixed

File: src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationWrrorReportServiceImpl.java (hunk 7 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationWrrorReportServiceImpl.java]: nope, it was a typo, thanks

**Comment by jhulslander** [comment]
Thanks for the catches, I believe I have addressed the items notes,  This is ready for review again.  Please also note there is a related PR in non-prod SQL

File: src/test/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationWrrorReportServiceImplTest.java (hunk 8 above)
Comment (jhulslander) [reply src/test/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationWrrorReportServiceImplTest.java]: fixed

File: src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationErrorReportServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+        }
+
+    }
+    
+    private String findPdpFromEmailAddress() {
+        return parameterService.getParameterValueAsString(KFSConstants.ParameterNamespaces.PDP, KfsParameterConstants.BATCH_COMPONENT, KFSConstants.FROM_EMAIL_ADDRESS_PARM_NM);
+    }
+    
+    private List<String> findPDPToEmailAddress() {
+        Collection addresses = parameterService.getParameterValuesAsString(KFSConstants.ParameterNamespaces.PDP, KfsParameterConstants.BATCH_COMPONENT,
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationErrorReportServiceImpl.java]: I shall switch this code to use CoreModuleNamespaces .... however me may want to create a technical debt user story to address the rest of our code base.

File: src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationErrorReportServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+    
+    private void initializeErrorReport() {
+        reportWriterService.setFileNamePrefix(configurationService.getPropertyValueAsString(CUPdpKeyConstants.PDP_SEND_ACH_NOTIFICATION_ERROR_REPORT_PREFIX));
+        reportWriterService.setTitle(configurationService.getPropertyValueAsString(CUPdpKeyConstants.PDP_SEND_ACH_NOTIFICATION_ERROR_REPORT_TITLE));
+        reportWriterService.initialize();
+        reportWriterService.writeNewLines(2);
+        String rowFormat = configurationService.getPropertyValueAsString(CUPdpKeyConstants.PDP_SEND_ACH_NOTIFICATION_ERROR_REPORT_FORMAT);
+        reportWriterService.writeFormattedMessageLine(rowFormat, "PAYEE ID", "PAYMENT GROUP", "DISBURSEMENT NUMBER", "EMAIL ADDRESS");
+        String hyphens25 = buildHyphenString(20);
+        String hyphens200 = buildHyphenString(50);
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationErrorReportServiceImpl.java]: good catch, I originally was going to use a longer string, I shall switch

File: src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationErrorReportServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+    
+    protected ReportWriterService reportWriterService;
+    protected ConfigurationService configurationService;
+    protected EmailService emailService;
+    protected ConcurBatchUtilityService concurBatchUtilityService;
+    protected ParameterService parameterService;
+
+    @Override
+    public File createBadEmailReport(List<PDPBadEmailRecord> badEmailRecords) {
+        LOG.info("createBadEmailReport, there are " + badEmailRecords.size() + " bad email records to report");
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationErrorReportServiceImpl.java]: sure, I shall change

**Comment by jhulslander** [comment]
I've made the requested changes, this is ready for review again

File: src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationErrorReportServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+            }
+            
+        } else {
+            LOG.error("emailBadEmailReport, no error report sent in");
+        }
+
+    }
+    
+    private String findPdpFromEmailAddress() {
+        return parameterService.getParameterValueAsString(KFSConstants.ParameterNamespaces.PDP, KfsParameterConstants.BATCH_COMPONENT, KFSConstants.FROM_EMAIL_ADDRESS_PARM_NM);
```
Comment (ajd299) [inline src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationErrorReportServiceImpl.java]: This one is still KFSContstants.ParameterNamespaces.PDP

Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuAchAdviceNotificationErrorReportServiceImpl.java]: opps

**Comment by jhulslander** [comment]
changes made, and ready for review again

**Comment by ajd299** [comment]
Looks good, will merge both PRs once unit tests pass

**Pull request merged**